Move AutogradMeta and DeviceGuardImplInterface virtual methods out-of-line.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #31899 Revert "Move AutogradMeta and DeviceGuardImplInterface virtual methods out-of-line."
* #31162 Don't use RTLD_GLOBAL to load _C.
* #31161 Uniformly apply Windows logic in cpp_extensions everywhere
* #31236 Don't unconditionally compile runJITCPPTests
* **#31176 Move AutogradMeta and DeviceGuardImplInterface virtual methods out-of-line.**
* #31157 Add missing TORCH_CUDA_API annotation to throw_nccl_error
* #31155 Remove dead CAFFE2_LIBS variable
* #31152 Remove LibIRC logic from cmake.

The goal is to make the typeinfos of these classes non-vague, so I don't
end up with a weak symbol that will get bound multiple times if I use
RTLD_LOCAL.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262580](https://our.internmc.facebook.com/intern/diff/D19262580)

ezyang committed:
Move AutogradMeta and DeviceGuardImplInterface virtual methods out-of-line.

The goal is to make the typeinfos of these classes non-vague, so I don't
end up with a weak symbol that will get bound multiple times if I use
RTLD_LOCAL.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 624d9df:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=624d9dfc82ac8d21828d8b8cd8f1b0a0bf31653e).

### 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3920372) pytorch_windows_build (1/1)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=73019067))
```
-- Generating C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/./torch_cuda_generated_THCTensorMathScan.cu.obj
C:/Users/circleci/project/build/win_tmp/bin/nvcc C:/Users/circleci/project/aten/src/THC/THCTensorMathScan.cu -c -o C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/./torch_cuda_generated_THCTensorMathScan.cu.obj -ccbin cl.exe -m64 -Dtorch_cuda_EXPORTS -DTORCH_CUDA_BUILD_MAIN_LIB -DTH_BLAS_MKL -D_OPENMP_NOFORCE_MANIFEST -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -D_CRT_SECURE_NO_DEPRECATE=1 -DWIN32_LEAN_AND_MEAN -DMAGMA_V2 -DAT_PARALLEL_OPENMP=1 -Xcompiler ,\"/DWIN32\",\"/D_WINDOWS\",\"/GR\",\"/w\",\"/MP\",\"/bigobj\",\"-openmp:experimental\",\"-DHAVE_AVX_CPU_DEFINITION\",\"-DHAVE_AVX2_CPU_DEFINITION\",\"/MD\",\"/O2\",\"/Ob2\",\"/w\",\"/MP\",\"/bigobj\" -Xcompiler /w -w -DONNX_NAMESPACE=onnx_torch -gencode arch=compute_75,code=sm_75 -Xcudafe --diag_suppress=cc_clobber_ignored -Xcudafe --diag_suppress=integer_sign_change -Xcudafe --diag_suppress=useless_using_declaration -Xcudafe --diag_suppress=set_but_not_used -Xcompiler -MD --expt-relaxed-constexpr --expt-extended-lambda -Xcompiler /wd4819 -Xcompiler /wd4503 -Xcompiler /wd4190 -Xcompiler /wd4244 -Xcompiler /wd4251 -Xcompiler /wd4275 -Xcompiler /wd4522 -Wno-deprecated-gpu-targets --expt-extended-lambda -gencode arch=compute_75,code=sm_75 -DCUDA_HAS_FP16=1 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ -DNVCC "-IC:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include" -IC:/Users/circleci/project/build/aten/src -IC:/Users/circleci/project/aten/src -IC:/Users/circleci/project/build -IC:/Users/circleci/project -IC:/Users/circleci/project/cmake/../third_party/googletest/googlemock/include -IC:/Users/circleci/project/cmake/../third_party/googletest/googletest/include -IC:/Users/circleci/project/third_party/protobuf/src -IC:/Users/circleci/project/build/win_tmp/mkl/include -IC:/Users/circleci/project/cmake/../third_party/benchmark/include -IC:/Users/circleci/project/cmake/../third_party/eigen -IC:/Jenkins/Miniconda3/include -IC:/Jenkins/Miniconda3/lib/site-packages/numpy/core/include -IC:/Users/circleci/project/cmake/../third_party/pybind11/include -I/opt/rocm/hip/include -I/include -IC:/Users/circleci/project/cmake/../third_party/cub -IC:/Users/circleci/project/build/caffe2/contrib/aten -IC:/Users/circleci/project/third_party/onnx -IC:/Users/circleci/project/build/third_party/onnx -IC:/Users/circleci/project/third_party/foxi -IC:/Users/circleci/project/build/third_party/foxi -IC:/Users/circleci/project/build/win_tmp/magma/include "-IC:/Program Files/NVIDIA Corporation/NvToolsExt/include" -IC:/Users/circleci/project/build/caffe2/aten/src/TH -IC:/Users/circleci/project/aten/src/TH -IC:/Users/circleci/project/build/caffe2/aten/src/THC -IC:/Users/circleci/project/aten/src/THC -IC:/Users/circleci/project/aten/src/THCUNN -IC:/Users/circleci/project/aten/src/ATen/cuda -IC:/Users/circleci/project/build/caffe2/aten/src -IC:/Users/circleci/project/aten/../third_party/catch/single_include -IC:/Users/circleci/project/aten/src/ATen/.. -IC:/Users/circleci/project/build/caffe2/aten/src/ATen -IC:/Users/circleci/project/c10/cuda/../.. -IC:/Users/circleci/project/c10/../ -IC:/Users/circleci/project/caffe2/../torch/csrc/api -IC:/Users/circleci/project/caffe2/../torch/csrc/api/include
THCTensorMathScan.cu
C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/detail/allocator/allocator_traits.inl(35): error C2993: 'T': illegal type for non-type template parameter '__formal'
C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/detail/allocator/allocator_traits.inl(35): note: see reference to class template instantiation 'thrust::detail::allocator_traits_detail::has_member_allocate_with_hint_impl_has_member<T,Result(Arg1,Arg2,Arg3)>' being compiled
C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/detail/allocator/allocator_traits.inl(35): error C2065: 't': undeclared identifier
C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/detail/allocator/allocator_traits.inl(35): error C2923: 'std::_Select<__formal>::_Apply': 't' is not a valid template type argument for parameter '<unnamed-symbol>'
C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/detail/allocator/allocator_traits.inl(35): error C2062: type 'unknown-type' unexpected
-- Removing C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/./torch_cuda_generated_THCTensorMathScan.cu.obj
"C:/Program Files/CMake/bin/cmake.exe" -E remove C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/./torch_cuda_generated_THCTensorMathScan.cu.obj
CMake Error at torch_cuda_generated_THCTensorMathScan.cu.obj.Release.cmake:281 (message):
  Error generating file
  C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/./torch_cuda_generated_THCTensorMathScan.cu.obj


[1968/2730] cmd.exe /C "cd /D C:\Users\circleci\project\build\caffe2\CMakeFiles\torch_cuda.dir\__\aten\src\THC\generated && "C:\Program Files\CMake\bin\cmake.exe" -E make_directory C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/. && "C:\Program Files\CMake\bin\cmake.exe" -D verbose:BOOL=ON -D build_configuration:STRING=Release -D generated_file:STRING=C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/./torch_cuda_generated_THCTensorSortByte.cu.obj -D generated_cubin_file:STRING=C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/./torch_cuda_generated_THCTensorSortByte.cu.obj.cubin.txt -P C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/torch_cuda_generated_THCTensorSortByte.cu.obj.Release.cmake"
-- Removing C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/./torch_cuda_generated_THCTensorSortByte.cu.obj
"C:/Program Files/CMake/bin/cmake.exe" -E remove C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/./torch_cuda_generated_THCTensorSortByte.cu.obj
-- Generating dependency file: C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/torch_cuda_generated_THCTensorSortByte.cu.obj.NVCC-depend
C:/Users/circleci/project/build/win_tmp/bin/nvcc -M -D__CUDACC__ C:/Users/circleci/project/aten/src/THC/generated/THCTensorSortByte.cu -o C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/torch_cuda_generated_THCTensorSortByte.cu.obj.NVCC-depend -ccbin cl.exe -m64 -Dtorch_cuda_EXPORTS -DTORCH_CUDA_BUILD_MAIN_LIB -DTH_BLAS_MKL -D_OPENMP_NOFORCE_MANIFEST -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -D_CRT_SECURE_NO_DEPRECATE=1 -DWIN32_LEAN_AND_MEAN -DMAGMA_V2 -DAT_PARALLEL_OPENMP=1 -Xcompiler ,\"/DWIN32\",\"/D_WINDOWS\",\"/GR\",\"/w\",\"/MP\",\"/bigobj\",\"-openmp:experimental\",\"-DHAVE_AVX_CPU_DEFINITION\",\"-DHAVE_AVX2_CPU_DEFINITION\",\"/MD\",\"/O2\",\"/Ob2\",\"/w\",\"/MP\",\"/bigobj\" -Xcompiler /w -w -DONNX_NAMESPACE=onnx_torch -gencode arch=compute_75,code=sm_75 -Xcudafe --diag_suppress=cc_clobber_ignored -Xcudafe --diag_suppress=integer_sign_change -Xcudafe --diag_suppress=useless_using_declaration -Xcudafe --diag_suppress=set_but_not_used -Xcompiler -MD --expt-relaxed-constexpr --expt-extended-lambda -Xcompiler /wd4819 -Xcompiler /wd4503 -Xcompiler /wd4190 -Xcompiler /wd4244 -Xcompiler /wd4251 -Xcompiler /wd4275 -Xcompiler /wd4522 -Wno-deprecated-gpu-targets --expt-extended-lambda -gencode arch=compute_75,code=sm_75 -DCUDA_HAS_FP16=1 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ -DNVCC "-IC:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include" -IC:/Users/circleci/project/build/aten/src -IC:/Users/circleci/project/aten/src -IC:/Users/circleci/project/build -IC:/Users/circleci/project -IC:/Users/circleci/project/cmake/../third_party/googletest/googlemock/include -IC:/Users/circleci/project/cmake/../third_party/googletest/googletest/include -IC:/Users/circleci/project/third_party/protobuf/src -IC:/Users/circleci/project/build/win_tmp/mkl/include -IC:/Users/circleci/project/cmake/../third_party/benchmark/include -IC:/Users/circleci/project/cmake/../third_party/eigen -IC:/Jenkins/Miniconda3/include -IC:/Jenkins/Miniconda3/lib/site-packages/numpy/core/include -IC:/Users/circleci/project/cmake/../third_party/pybind11/include -I/opt/rocm/hip/include -I/include -IC:/Users/circleci/project/cmake/../third_party/cub -IC:/Users/circleci/project/build/caffe2/contrib/aten -IC:/Users/circleci/project/third_party/onnx -IC:/Users/circleci/project/build/third_party/onnx -IC:/Users/circleci/project/third_party/foxi -IC:/Users/circleci/project/build/third_party/foxi -IC:/Users/circleci/project/build/win_tmp/magma/include "-IC:/Program Files/NVIDIA Corporation/NvToolsExt/include" -IC:/Users/circleci/project/build/caffe2/aten/src/TH -IC:/Users/circleci/project/aten/src/TH -IC:/Users/circleci/project/build/caffe2/aten/src/THC -IC:/Users/circleci/project/aten/src/THC -IC:/Users/circleci/project/aten/src/THCUNN -IC:/Users/circleci/project/aten/src/ATen/cuda -IC:/Users/circleci/project/build/caffe2/aten/src -IC:/Users/circleci/project/aten/../third_party/catch/single_include -IC:/Users/circleci/project/aten/src/ATen/.. -IC:/Users/circleci/project/build/caffe2/aten/src/ATen -IC:/Users/circleci/project/c10/cuda/../.. -IC:/Users/circleci/project/c10/../ -IC:/Users/circleci/project/caffe2/../torch/csrc/api -IC:/Users/circleci/project/caffe2/../torch/csrc/api/include
THCTensorSortByte.cu
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

ezyang committed:
Update on "Move AutogradMeta and DeviceGuardImplInterface virtual methods out-of-line."


The goal is to make the typeinfos of these classes non-vague, so I don't
end up with a weak symbol that will get bound multiple times if I use
RTLD_LOCAL.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Move AutogradMeta and DeviceGuardImplInterface virtual methods out-of-line."


The goal is to make the typeinfos of these classes non-vague, so I don't
end up with a weak symbol that will get bound multiple times if I use
RTLD_LOCAL.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

albanD reviewed (approved):
Looks good.

ezyang committed:
Update on "Move AutogradMeta and DeviceGuardImplInterface virtual methods out-of-line."


The goal is to make the typeinfos of these classes non-vague, so I don't
end up with a weak symbol that will get bound multiple times if I use
RTLD_LOCAL.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19262580](https://our.internmc.facebook.com/intern/diff/D19262580)

[ghstack-poisoned]
